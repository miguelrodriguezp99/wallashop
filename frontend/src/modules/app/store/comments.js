import { create } from "zustand";
import { getPostComments } from "../../../backend/postService";

export const useCommentsStore = create((set, get) => ({
	commentsByPost: [],

	fetchComments: async (postId) => {
		try {
			getPostComments(
				postId,
				(data) => {
					console.log("data", data);
					// Compara el nuevo array de comentarios con el existente
					if (!arraysEqual(data, get().commentsByPost)) {
						set({ commentsByPost: data });
					}
				},
				(errors) => {
					console.log(errors);
				},
			);
			// onSuccess y OnErrors están metidos en la propia función en este caso
		} catch (error) {
			console.error(error);
		}
	},
}));

// Función para comparar dos arrays
function arraysEqual(arr1, arr2) {
	return JSON.stringify(arr1) === JSON.stringify(arr2);
}
