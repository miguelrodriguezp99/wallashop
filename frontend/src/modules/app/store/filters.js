import { create } from "zustand";
import { getPosts } from "../../../backend/postService";

export const useFiltersStore = create((set, get) => ({
	price: "0",
	category: "all",
	type: "all",
	query: "",
	posts: [],
	filteredPosts: [],
	stillValid: null,
	created: null,
	expDate: null,
	sort: null,

	setStillValid: (stillValid) => set({ stillValid }),
	setCreated: (created) => set({ created }),
	setExpDate: (expDate) => set({ expDate }),

	setPriceFilter: (price) => set({ price }),

	setCategoryFilter: (category) => set({ category }),

	setTypeFilter: (type) => set({ type }),

	setPosts: (posts) => set({ posts }),

	setFilteredPosts: (filteredPosts) => set({ filteredPosts }),

	setQuery: (query) => set({ query }),

	setSort: (sort) => set({ sort }),

	getFilters: () => ({
		price: get().price,
		category: get().category,
	}),

	fetchPosts: async () => {
		try {
			getPosts(
				{ page: 0, size: 10 },
				(data) => {
					set({ posts: data.content });
				},
				(errors) => {
					console.log(errors);
				},
			);
		} catch (error) {
			console.error(error);
		}
	},

	filterPosts: () => {
		const {
			price,
			category,
			type,
			query,
			posts,
			stillValid,
			created,
			expDate,
		} = get();

		const filteredPosts1 = posts.filter((post) => {
			const meetsPriceCriteria = post.price >= parseFloat(price);
			const meetsCategoryCriteria =
				category === "all" || post.category.name === category;
			const meetsTypeCriteria = type === "all" || post.type === type;
			const meetsSearchCriteria =
				post.title.toLowerCase().includes(query.toLowerCase()) ||
				post.description.toLowerCase().includes(query.toLowerCase());

			let meetsStillValidCriteria = true;
			if (stillValid !== null || isNaN(stillValid)) {
				meetsStillValidCriteria =
					(post.stillValid !== null || post.stillValid !== undefined) &&
					Date.parse(post.stillValid) >= stillValid;
			}

			let meetsCreatedCriteria = true;
			if (created !== null || isNaN(created)) {
				meetsCreatedCriteria =
					(post.creationDate !== null || post.creationDate !== undefined) &&
					Date.parse(post.creationDate) >= created;
			}

			let meetsExpiratedCriteria = true;
			if (expDate !== null || isNaN(expDate)) {
				meetsExpiratedCriteria =
					(post.expirationDate !== null || post.expirationDate !== undefined) &&
					Date.parse(post.expirationDate) >= expDate;
			}

			return (
				meetsPriceCriteria &&
				meetsCategoryCriteria &&
				meetsTypeCriteria &&
				meetsSearchCriteria &&
				meetsStillValidCriteria &&
				meetsCreatedCriteria &&
				meetsExpiratedCriteria
			);
		});

		const filteredPosts = [...filteredPosts1];
		const { sort } = get();
		switch (sort) {
			case 'NameAsc':
			  filteredPosts.sort((a, b) => a.title.localeCompare(b.title));
			  break;
			case 'titleDesc':
			  filteredPosts.sort((a, b) => b.title.localeCompare(a.title));
			  break;
			case 'CategoryAsc':
			  filteredPosts.sort((a, b) => a.category.name.localeCompare(b.category.name));
			  break;
			case 'CategoryDesc':
			  filteredPosts.sort((a, b) => b.category.name.localeCompare(a.category.name));
			  break;
			case 'TypeAsc':
			  filteredPosts.sort((a, b) => a.type.localeCompare(b.type));
			  break;
			case 'TypeDesc':
			  filteredPosts.sort((a, b) => b.type.localeCompare(a.type));
			  break;
			case 'PriceAsc':
			  filteredPosts.sort((a, b) => a.price - b.price);
			  break;
			case 'PriceDesc':
			  filteredPosts.sort((a, b) => b.price - a.price);
			  break;
			case 'ValidAsc':
				filteredPosts.sort((a, b) => {
					if (a.stillValid != null && b.stillValid != null) return a.stillValid.localeCompare(b.stillValid);
					if (a.stillValid != null) return -1;
					else if (b.stillValid != null) return -1;
					return 0;
				  });
			  break;
			case 'ValidDesc':
				filteredPosts.sort((a, b) => {
					if (a.stillValid != null && b.stillValid != null) return a.stillValid.localeCompare(b.stillValid);
					if (a.stillValid != null) return -1;
					else if (b.stillValid != null) return -1;
					return 0;
				  });
				  break;
			case 'CreatedAsc':
			  filteredPosts.sort((a, b) => a.creationDate.localeCompare(b.creationDate));
			  break;
			case 'CreatedDesc':
			  filteredPosts.sort((a, b) => b.creationDate.localeCompare(a.creationDate));
			  break;
			case 'ExpiresAsc':
			  filteredPosts.sort((a, b) => a.expirationDate.localeCompare(b.expirationDate));
			  break;
			case 'ExpiresDesc':
			  filteredPosts.sort((a, b) => b.expirationDate.localeCompare(a.expirationDate));
			  break;
			default:
			  break;
		}

		set({ filteredPosts });
	},

	getPostById: (id) => {
		const { posts } = get();

		if (posts.length === 0) {
			get().fetchPosts();
			return null;
		}

		return posts.find((post) => post.id === id);
	},
}));
