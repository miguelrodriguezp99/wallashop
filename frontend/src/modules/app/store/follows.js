import { create } from "zustand";
import { getPostFollows } from "../../../backend/postService";

export const useFollowsStore = create((set, get) => ({
	followsByPost: [],

	fetchFollows: async (postId) => {
		try {
			getPostFollows(
				postId,
				(data) => {
					console.log("data", data);
					// Compara el nuevo array de follows con el existente
					if (!arraysEqual(data, get().followsByPost)) {
						set({ followsByPost: data });
					}
				},
				(errors) => {
					console.log(errors);
				},
			);
			// onSuccess y OnErrors están metidos en la propia función en este caso
		} catch (error) {
			console.error(error);
		}
	},
}));

// Función para comparar dos arrays
function arraysEqual(arr1, arr2) {
	return JSON.stringify(arr1) === JSON.stringify(arr2);
}