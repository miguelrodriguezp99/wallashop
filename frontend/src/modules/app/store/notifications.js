import { create } from "zustand";
import { getNotifications, deleteNotification } from "../../../backend/userService";

export const useNotificationsStore = create((set, get) => ({
  notificationsByUser: [],

  fetchNotifications: async (userId) => {
    try {
      getNotifications(
        userId,
        (data) => {
          const notificationsWithId = data.map((notification) => ({
            ...notification,
            id: notification.id
          }));
          set({ notificationsByUser: notificationsWithId });
        },
        (errors) => {
          console.log(errors);
        },
      );
    } catch (error) {
      console.error(error);
    }
  },
  
  deleteNotification: async (userId, notificationId) => {
    try {
      deleteNotification(
        userId,
        notificationId,
        () => {
          console.log('Notificación eliminada con éxito');
          set((state) => ({
            notificationsByUser: state.notificationsByUser.filter(
              (notification) => notification.id !== notificationId
            ),
          }));
        },
        (errors) => {
          console.error('Error al eliminar la notificación:', errors);
        }
      );
    } catch (error) {
      console.error(error);
    }
  },

}));
