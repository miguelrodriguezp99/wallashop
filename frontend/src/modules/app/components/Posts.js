import { useFiltersStore } from "../store/filters";
/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect } from "react";

import "../styles/posts.css";

import PostCard from "./PostCard";

const Posts = () => {
	const {
		price,
		category,
		type,
		fetchPosts,
		filterPosts,
		posts,
		filteredPosts,
		query,
		stillValid,
		created,
		expDate,
		sort
	} = useFiltersStore();

	useEffect(() => {
		fetchPosts();
	}, []);

	// ---- AVOID FETCH POSTS ----
	//const postsJSON = PostsJSON.posts;

	useEffect(() => {
		filterPosts();
	}, [price, category, posts, type, query, stillValid, created, expDate, sort]);

	return (
		<>
			<section className="container mx-auto p-10 md:py-20 px-0 md:p-10 md:px-0 bg-black mt-5">
				<section className="p-5 md:p-0 grid grid-cols-1 sm:grid-cols-1 md:grid-cols-2 xl:grid-cols-3 2xl:grid-cols-4 gap-5 items-start">
					{filteredPosts.map((post, index) => {
						return <PostCard key={index} post={post} />;
					})}
				</section>
			</section>
		</>
	);
};

export default Posts;
