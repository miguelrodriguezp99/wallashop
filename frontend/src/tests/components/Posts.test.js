import React from "react";
import renderer from "react-test-renderer";
import Posts from "../../modules/app/components/Posts";

jest.mock("../../modules/app/store/filters", () => ({
  useFiltersStore: jest.fn(),
}));

describe("Posts Component", () => {
  it("renders correctly", () => {
    const mockUseFiltersStore = {
      price: 0,
      category: "",
      type: "",
      fetchPosts: jest.fn(),
      filterPosts: jest.fn(),
      posts: [],
      filteredPosts: [],
      query: "",
      stillValid: true,
      created: "",
      expDate: "",
    };

    require("../../modules/app/store/filters").useFiltersStore.mockReturnValue(mockUseFiltersStore);
    const component = renderer.create(<Posts />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});