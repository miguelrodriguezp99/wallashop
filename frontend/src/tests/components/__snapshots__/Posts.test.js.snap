// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts Component renders correctly 1`] = `
<section
  className="container mx-auto p-10 md:py-20 px-0 md:p-10 md:px-0 bg-black mt-5"
>
  <section
    className="p-5 md:p-0 grid grid-cols-1 sm:grid-cols-1 md:grid-cols-2 xl:grid-cols-3 2xl:grid-cols-4 gap-5 items-start"
  />
</section>
`;
